ark/phenotyping/som_utils.py

```diff
@@ -211,27 +330,43 @@ def compute_cell_cluster_channel_avg(pixel_channel_avg, cell_counts,
 
     # verify the pixel_cluster_col provided is valid
     misc_utils.verify_in_list(
-        provided_cluster_col=cluster_col,
-        valid_cluster_cols=['cluster', 'hCluster_cap']
+        provided_cluster_col=[pixel_cluster_col],
+        valid_cluster_cols=['pixel_som_cluster', 'pixel_meta_cluster']
     )
 
     # subset over the provided fovs
     cell_counts_sub = cell_counts[cell_counts['fov'].isin(fovs)].copy()
 
     # subset over the cluster count columns of pixel_channel_avg
-    cluster_cols = [c for c in cell_counts_sub.columns.values if cluster_col in c]
+    cluster_cols = [c for c in cell_counts_sub.columns.values if pixel_cluster_col in c]
     cell_counts_clusters = cell_counts_sub[cluster_cols].copy()
 
     # sort the columns of cell_counts_clusters in ascending cluster order
+    # sort by int value, not string value
     cell_counts_clusters = cell_counts_clusters.reindex(
-        sorted(cell_counts_clusters.columns.values), axis=1
+        sorted(cell_counts_clusters.columns.values,
+               key=lambda x: int(x.replace(pixel_cluster_col + '_', ''))),
+        axis=1
     )
 
-    # sort the pixel_channel_avg table by cluster_col in ascending cluster order
-    pixel_channel_avg_sorted = pixel_channel_avg.sort_values(by=cluster_col)
+    # sort the pixel_channel_avg table by pixel_cluster_col in ascending cluster order
+    pixel_channel_avg_sorted = pixel_channel_avg.sort_values(by=pixel_cluster_col)
+
+    # check that the same clusters are in both cell_counts_clusters and pixel_channel_avg_sorted
+    # the matrix multiplication will fail if this is not caught
+    cell_counts_cluster_ids = [
+        int(x.replace(pixel_cluster_col + '_', '')) for x in cell_counts_clusters.columns.values
+    ]
+    pixel_channel_cluster_ids = pixel_channel_avg[pixel_cluster_col].values
+
+    misc_utils.verify_same_elements(
+        enforce_order=True,
+        cell_counts_cluster_ids=cell_counts_cluster_ids,
+        pixel_channel_cluster_ids=pixel_channel_cluster_ids
+    )
 
     # subset over just the markers of pixel_channel_avg
-    pixel_channel_avg_sub = pixel_channel_avg_sorted.drop(columns=cluster_col)
+    pixel_channel_avg_sub = pixel_channel_avg_sorted.drop(columns=pixel_cluster_col)
```

Comment:
Instead of dropping columns here, you should directly subset for the correct column names. When a user chooses to do cell clustering using "pixel_som_cluster", pixel_channel_avg contains the columns "pixel_meta_cluster" as well as "count." Since these columns are not explicitly dropped, they persist in the final weighted data table and have weird values because of the matrix multiplication.